composer/trainer/trainer.py

```diff
@@ -2204,6 +2255,41 @@ def eval(
         if original_num_batches is not None:
             self.state.dataloader_len = original_num_batches
 
+    def _eval_batch(self, data_spec: DataSpec, metrics: MetricCollection):
```

Comment:
prefer to keep this in the `eval` method if possible, so that all the events are shown linearly in the eval loop

Reply:
can you elaborate? this is the same decomposition as the train loop -- I prefer at least some decomposition over one massive function

Reply:
The train loop structure was done more for code readability/reuse, but never purely for decomposition:
* `_train_batch:` prevent excessive indentation (while, try/catch, etc.)
* `_train_microbatches`: multiple calls depending on closures/etc.
* `_train_microbatch`: prevent excessive indentation (using sync contexts)

As `_eval_batch` is not that long block of code anyway, prefer not to disrupt the readability of EVAL_* events in the code + add another layer in the stack trace.